src/pose_estimating.h

```diff
@@ -0,0 +1,162 @@
+namespace PoseEstimating {
+
+/**
+ * @brief An enum for each important pose body part
+ */
+enum Joint {
+    Head,
+    Neck,
+    Shoulder,
+    Hip,
+    Knee,
+    Foot
+}
+
+/*
+ * @brief A representation of a body part
+ *
+ * Each body part is represented as a `Pose::ConnectedJoint` to record the joints relative position and the connected joints to this joint
+ * 
+ */
+struct ConnectedJoint {
+    Joint joint;
+    /* @brief relative x co-ordinate of joint */
+    float x;
+    /* @brief relative y co-ordinate of joint */
+    float y;
+    ConnectedJoint * upper_connected_joint;
+    /* 
+     * @brief Angle(in degrees) of upper joint measured anticlockwise from bottom left corner  
+     */
+    float upper_angle;
+    ConnectedJoint * lower_connected_joint;
+    /* 
+     * @brief Angle(in degrees) of lower joint measured anticlockwise from bottom left corner  
+     */
+    float lower_angle;
+}
+
+/* 
+ *  @brief The `Pose::Pose` representation of a human's pose, containing all the expected `Pose::ConnectedJoint 
+ */
+struct Pose {
+    ConnectedJoint head;
+    ConnectedJoint neck;
+    ConnectedJoint shoulder;
+    ConnectedJoint hip;
+    ConnectedJoint knee;
+    ConnectedJoint foot;
+}
+
+/*
+ * @brief This class handles representations of the user's pose and calculates any updates to their pose that is required.
+ * 
+ * This class stores representations of the users current pose and their calibrated ideal pose, from the results of the PostProcessor class.
+ *
+ * This class handles:
+ * Updatin current user's pose or ideal user's pose
+ * Calculating the angle between ConnectedJoints
+ * Comparing the current user's pose and ideal user's pose
+ * 
+ */
+class PoseEstimater {
+ public:
+     /*
+      * @brief The user's current pose from most recent data provided
+     */
+     Pose currentPose;
+     /* 
+      * @brief The user's ideal pose, expected to be generated during user calibration  
+      */
+     Pose idealPose;
+     /*
+      * @brief Returns the user's current pose
+      * 
+      * @returns Pose
+      */
+     Pose get_current_pose();
+     /*
+      * @brief Updates the user's current pose from PostProcesser results
+      *
+      * @param results ProcessedResults struct containing user's pose data
+      */
+     void updateCurrentPose(PostProcessing :: ProcessedResults results);
+     /*
+      * @brief Returns the user's ideal pose
+      * 
+      * @returns Pose
+      */
+     Pose get_ideal_pose();
+     /*
+      * @brief Updates the user's ideal pose from PostProcesser results
+      *
+      * @param results ProcessedResults struct containing user's pose data
+      */
+     void updateIdealPose(PostProcessing :: ProcessedResults results);
+     /*
+      * @brief Compares user's current pose and ideal pose, returing a pose object, with the angles of the ConnectJoints representing the change needed for each one
+      *
+      * @returns Pose
+      */  
+     Pose calculatePoseChangeNeeded();
```

Comment:
How should this class be used? Will there be two function calls for every frame output by the post processor? It would be good if there is a single `run()` or similar. We can then use different getters to get the different outputs you have suggested (like current pose and ideal pose and difference to ideal pose).

Reply:
I can add a new run function, but I feel both of these should still be public

Reply:
Are we expecting to check the posture on every frame? And then what metric do we use to decided they've been in a bad posture for too long? I kinda imagined you'd check the bad posture every so often? If they're in a good posture, only check in a few mins time etc? But regardless I've put in a function updateCheckCurrentPose()

Reply:
I guess that'll be the responsibility of that first interface with the camera - when we come to implementing that, we would want it to be able to capture frames at various rates and potentially change the rate depending on if a good posture has been detected etc. I don't think any of that should affect the implementation of this component though?

Reply:
It justifies keeping having a separate update and check function alongside an update and check function

Reply:
I don't quite understand what you mean by that last comment?

Reply:
So we can have just a function that updates the users posture and checks the users posture in one go. This is what miklas is suggestion with the run() function. However, I can see the use case where we want the live posture checker to run updates on what the user's pose looks like but may only want to check the if the posture is bad every few minutes. But I can also see checking the posture alongside updating the posture. So my proposal is to have a function for updating the users posture, another for checking if the user's posture is good and another which does both

Reply:
What I meant was more that there should probably only be a single setter function that is called when a new prediction of current posture is available. We would then be free to call any of the getters that format the pose in different ways: current pose, difference of current pose to ideal, etc.

There is a danger/difficulty when multiple setters exist of getting confused between them when using the interface. It is also more difficult to update the class as changes in one setter function may need to be copied to the other setters as well.

---

I may have initially misunderstood the purpose of the different public methods, but I think I may have a better understanding now. Is the idea that one can?:
- Tune the class to a particular ideal posture with `updateIdealPose()`
- Pass in the real current pose with `updateCurrentPose()`
- Get the amount by which a user needs to adjust their pose with `calculatePoseChangeNeeded()`

If that is the case, could you update the documentation to be a bit more explicit?

Reply:
I may have added too many functions. I took the approach the user of this class may have their own uses for it outside of what we envison, which is why I've added various combinations as opposed to just a one "run" function. However, I feel now, especially with my recent update, I've gotten carried away. I think we may need an "in person" discussion

Reply:
We've discussed this and the updated code looks good; a lot lighter and clearer